### PR TITLE
Issue #5969: Respect and expect tab width in column number check for lambda indentations

### LINKED ISSUE #5969: Indentation violation for lambda depends on tab symbols in code
ATTENTION: code with tab symbols.
```
/tmp $ javac Test.java

/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
		"-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
		"https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <module name="TreeWalker">
        <property name="tabWidth" value="4"/>
        <module name="Indentation"/>
    </module>
</module>

/tmp $ cat Test.java
import java.lang.*;
import java.util.*;
import java.util.stream.*;

class Test {
	List<Integer> test(List<String> input) {
		return input.stream()
				.flatMap(in -> Arrays.stream(in.split(""))) // only 2 char parameter has violation
				.map(Integer::valueOf)
				.collect(Collectors.toList());
	}

	String test2(String input) {
		return Optional.ofNullable(input)
				.filter(inp -> inp.contains("e")) // only 3 char parameter has violation
				.orElse(null);
	}
}

/tmp $ java -jar checkstyle-8.10.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /tmp/Test.java:8: 'lambda' has incorrect indentation level 16, expected level should be 20. [Indentation]
[ERROR] /tmp/Test.java:15: 'lambda' has incorrect indentation level 16, expected level should be 20. [Indentation]
Audit done.
Checkstyle ends with 2 errors.
```

---------------

Expectation: lambda parameter name length should not affect the indentation rule, and the example code above should not have any violations.

If you change line 8's lamdba parameter to `i` or `inp` or any other name that is not 2 characters, the indentation violation goes away.
Likewise, line 15's lambda parameter can be changed to any name that is not 3 characters and the indentation violation goes away.

--------------

If tabs changed to spaces, no problems are reported, but we use tabs in code.
http://checkstyle.sourceforge.net/config_misc.html#Indentation no special property for tab size, tab should be taken from Treewalker.

### LINKED ISSUE #7675: Confusing violation of Indentation rule
Im having a problem with the https://checkstyle.sourceforge.io/config_misc.html#Indentation check.

```/home/serra/dev/cs-repro $ javac Test.java```
[no output here]
```/home/serra/dev/cs-repro $ cat my-checks.xml```
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
	<property name="charset" value="UTF-8" />
	<property name="tabWidth" value="3" />
	<module name="TreeWalker">
		<module name="Indentation">
			<property name="basicOffset" value="3" />
			<property name="caseIndent" value="0" />
			<property name="lineWrappingIndentation" value="6" />
		</module>
	</module>
</module>
```

```/home/serra/dev/cs-repro $ cat Test.java```
```java
import java.util.Arrays;
import java.util.stream.Stream;

public class Test {
	public void test(Stream<Inner<?>> stream) {
		stream
				.filter(ps -> ps instanceof Inner)
				.map(ps -> ((Inner<?>) ps).getPropertyNames())
				.forEach(System.out::println);
	}

	private static class Inner<T> {
		String[] getPropertyNames() {
			return new String[] {"a", "b"};
		}
	}
}
```
```/home/serra/dev/cs-repro $ java -Duser.language=en -Duser.country=us -jar checkstyle-8.29-all.jar -c my-checks.xml Test.java
Starting audit...
[ERROR] /home/serra/dev/cs-repro/Test.java:8: 'lambda' has incorrect indentation level 12, expected level should be 18. [Indentation]
Audit done.
```

---------------

I expected the code to be conforming to my rules.
As far as I can recognize I adhere to all the rules in the xml.

I am not sure if this is a bug or if I am just misunderstandung how the rule is working.

### PR DESCRIPTION
**Pull request by liach**
…lambda indentations

Fixes #5969 and closes #7675

I just hope to live a happy life of Minecraft modding. Thank you for this great checkstyle! https://github.com/FabricMC/fabric/pull/518/files#diff-23a850fe9b00699c6f6cd295e8ef341d

### CODE REVIEW

**Comment by rnveach** [comment]
We will also require regression since this is a change to production code.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ protected IndentLevel getIndentImpl() {
         // If the start of the lambda is the first element on the line;
         // assume line wrapping with respect to its parent.
         final DetailAST firstChild = getMainAst().getFirstChild();
-        if (getLineStart(firstChild) == firstChild.getColumnNo()) {
+        if (getLineStart(firstChild) == expandedTabsColumnNo(firstChild)) {
```
[the remark this answers is not in the document]
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java]: Ignore this. The left is getting the start of the line and ensuring it matches.

[the remark this answers is not in the document]
Comment (liach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java]: Imo your analysis is wrong. `getLineStart` returns a tab expanded column number while `firstChild.getColumnNo` counts tab as single-width.

**Comment by liach** [comment]
Updated. Just curious, what is a regression test? like https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-tester or the jacoco report?

**Comment by rnveach** [comment]
Regression is https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-tester so we can see how your changes impacts other projects.

Indentation is also a very special issue with regression. You may need to customize the check for each project to ensure the violations produced are accurate. See my files in https://github.com/rnveach/contribution/tree/rvmaster2/checkstyle-tester matching "my_checks_XXX.xml`.

**Comment by liach** [comment]
@rnveach Sorry to bother, but unfortunately I failed to get `diff.groovy` running.
```
ERROR: JAVA_HOME is set to an invalid directory: C:/Program Files/Java/jdk1.8.0_211
Please set the JAVA_HOME variable in your environment
to match the location of your Java installation.
Caught: groovy.lang.GroovyRuntimeException: Error: !
groovy.lang.GroovyRuntimeException: Error: !
        at diff.executeCmd(diff.groovy:353)
        at diff.executeCmd(diff.groovy:347)
        at diff$executeCmd$9.callCurrent(Unknown Source)
        at diff.generateCheckstyleReport(diff.groovy:187)
        at diff$generateCheckstyleReport$5.callCurrent(Unknown Source)
        at diff.run(diff.groovy:26)
```
I am sure that there is a valid jdk installed at `C:/Program Files/Java/jdk1.8.0_211`. I have tried using other jdk home variables, and have maven, gradle, and groovy available on `PATH` environmental variable. I want to generate a diff report and do need some help here.

**Comment by rnveach** [comment]
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/diff.groovy#L187
Error happens when executing `groovy`.

Are you able to execute groovy outside the script via `groovy launch.groovy`? 
Can you show us the output of `where java` and `echo %JAVA_HOME%` in the command prompt.

**Comment by liach** [comment]
Command prompt (Windows cmd)
```
Microsoft Windows [Version 10.0.18363.720]
(c) 2019 Microsoft Corporation. All rights reserved.

C:\code\checkstyle-contribution\checkstyle-tester>where java
where java
C:\Program Files\Amazon Corretto\jdk11.0.6_10\bin\java.exe
C:\Program Files\AdoptOpenJDK\jdk-11.0.6.10-openj9\bin\java.exe
C:\Program Files\Amazon Corretto\jdk1.8.0_212\bin\java.exe
C:\Program Files\AdoptOpenJDK\jdk-8.0.202.08\bin\java.exe
C:\Program Files\AdoptOpenJDK\jre-8.0.202.08\bin\java.exe
C:\Program Files\Java\jdk1.8.0_202\bin\java.exe

C:\code\checkstyle-contribution\checkstyle-tester>echo %JAVA_HOME%
echo %JAVA_HOME%
C:\Program Files\Java\jdk1.8.0_211

C:\code\checkstyle-contribution\checkstyle-tester>

```
In powershell or git bash (unix bash, where I execute the groovy script from), the `echo %JAVA_HOME%` prints `%JAVA_HOME%` instead. Now that I switched to command prompt, it seems to work fine! Thanks.

**Comment by rnveach** [comment]
Look at the results of where and echo. You Java home is not set to the right directory.
> C:\Program Files\Java\jdk1.8.0_202\bin\java.exe
> C:\Program Files\Java\jdk1.8.0_211

Unless you have a JDK that where isn't finding, it looks like it is pointing to the wrong directory.

**Comment by liach** [comment]
[Reports diff](https://liachmodded.github.io/checkstyle/fix/lambda-tab/reports/diff/) and [Jacoco](https://liachmodded.github.io/checkstyle/fix/lambda-tab/jacoco/) published.

The full report is like 5GB so I only published the diff. Openjdk wasn't included because checkstyle uses too much memory, which my device cannot support.

The `InputIndentationLambda3` has included failure cases in the last push. The travis build may need a restart, as it seems to fail for bad network connection.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java [hunk 2]
```diff
@@ -1759,6 +1759,16 @@ public void testLambda2() throws Exception {
...
     }
 
+    @Test
+    public void testLambda3() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(IndentationCheck.class);
+        checkConfig.addAttribute("tabWidth", "4");
+        checkConfig.addAttribute("basicOffset", "4");
+        checkConfig.addAttribute("lineWrappingIndentation", "8");
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWarns(checkConfig, getPath("InputIndentationLambda3.java"), expected);
```
[the remark this answers is not in the document]
Comment (liach) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: Already done.

[the remark this answers is not in the document]
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: What happened to the file with no violations? We should have both good cases and bad cases. Also ensure that the case from the issue is part of the UTs.

[the remark this answers is not in the document]
Comment (liach) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: Added back, waiting for ci for now.

[the remark this answers is not in the document]
Comment (liach) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java]: Done, ci build now passes.

**Comment by liach** [comment]
@rnveach Updated. Mind review again?

**Comment by pbludov** [comment]
@liach one minor:
The Spring Framework project uses tabs for [indentation](https://github.com/spring-projects/spring-framework/wiki/Code-Style#indentation). Please rerun the report with this config: 
```xml
    <module name="TreeWalker">
        <property name="tabWidth" value="2"/>
         <module name="Indentation">
             <property name="throwsIndent" value="2"/>
             <property name="arrayInitIndent" value="2"/>
             <property name="caseIndent" value="0"/>
             <property name="basicOffset" value="2"/>
             <property name="lineWrappingIndentation" value="4"/>
         </module>
  </module>
```

`spring-framework` and `spoon` only.

**Comment by liach** [comment]
@pbludov Reports updated, still at https://liachmodded.github.io/checkstyle/fix/lambda-tab/reports/diff/
If you looked at the old report, might need to refresh page in your browser

**Comment by liach** [comment]
Thank you very much! :heart:

**Pull request merged**